### PR TITLE
Feature/mqtt light

### PR DESCRIPTION
**Pull request by harmjanr**

### CODE REVIEW

File: WordClock/Mqtt.cpp [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+  // Concat the payload into a string
+  String payload;
+  for (uint8_t i = 0; i < p_length; i++) {
+    payload.concat((char)p_payload[i]);
+  }
+
+  Serial.println(p_topic);
+  Serial.println(payload);
+  
+//  // Handle message topic
```
Comment (jortberends) [inline WordClock/Mqtt.cpp]: //  //

File: WordClock/Mqtt.cpp [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+      Serial.println("INFO: connected");
+      // Once connected, publish an announcement...
+      publishLightState();
+      // ... and resubscribe
+      client.subscribe(MQTT_LIGHT_RGB_COMMAND_TOPIC);
+      client.subscribe(MQTT_LIGHT_BRIGHTNESS_COMMAND_TOPIC);
+    } else {
+      Serial.print("ERROR: failed, rc=");
+      Serial.print(client.state());
+      Serial.println("DEBUG: try again in 5 seconds");
```
Comment (jortberends) [inline WordClock/Mqtt.cpp]: 5 seconds? i think it's retrying immediately

File: WordClock/Mqtt.cpp [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    setRgbState(payload);
+  } else if (String(MQTT_LIGHT_BRIGHTNESS_COMMAND_TOPIC).equals(p_topic)) { 
+    setBrightness(payload);
+  }
+
+}
+
+void reconnect() {
+  // Loop until we're reconnected
+  while (!client.connected()) {
```
Comment (jortberends) [inline WordClock/Mqtt.cpp]: I think we should only try to reconnect once every loop. If the MQQT server goes down, the clock will stay in this loop until it's reconnected and not update the clock in the mean time.

File: WordClock/Mqtt.cpp [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+WiFiClient wifiClient;
+PubSubClient client(wifiClient);
+
+Display* _display;
+
+Mqtt::Mqtt() {
+}
+
+void setRgbState(String payload) {
```
Comment (jortberends) [inline WordClock/Mqtt.cpp]: This func is not defined in the h file

File: WordClock/Mqtt.cpp [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+  String b2 = payload.substring(6);
+
+  int r = payload.substring(0,3).toInt();
+  int g = payload.substring(3,6).toInt();
+  int b = payload.substring(6).toInt();
+
+  _display->setColor(r, g, b);
+}
+
+void setBrightness(String payload) {
```
Comment (jortberends) [inline WordClock/Mqtt.cpp]: This func is not defined in the h file

File: WordClock/Mqtt.cpp [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+  int brightness = payload.toInt();
+  _display->setBrightness(brightness);
+}
+
+// function called to publish the state of the led (on/off)
+void publishLightState() {
```
Comment (jortberends) [inline WordClock/Mqtt.cpp]: This func is not defined in the h file

File: WordClock/Mqtt.cpp [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+    client.publish(MQTT_LIGHT_STATE_TOPIC, "on", true);
+}
+
+// Function called when a MQTT message arrived
+void callback(char* p_topic, byte* p_payload, unsigned int p_length) {
```
Comment (jortberends) [inline WordClock/Mqtt.cpp]: This func is not defined in the h file

File: WordClock/Mqtt.cpp [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+//  // Handle message topic
+  if (String(MQTT_LIGHT_RGB_COMMAND_TOPIC).equals(p_topic)) { 
+    setRgbState(payload);
+  } else if (String(MQTT_LIGHT_BRIGHTNESS_COMMAND_TOPIC).equals(p_topic)) { 
+    setBrightness(payload);
+  }
+
+}
+
+void reconnect() {
```
Comment (jortberends) [inline WordClock/Mqtt.cpp]: This func is not defined in the h file

Comment (jortberends) [reply WordClock/Mqtt.cpp]: And not a member of the mqtt class

File: WordClock/Mqtt.cpp (hunk 7 above)
Comment (jortberends) [reply WordClock/Mqtt.cpp]: And not a member of the mqtt class

File: WordClock/Mqtt.cpp (hunk 6 above)
Comment (jortberends) [reply WordClock/Mqtt.cpp]: And not a member of the mqtt class

File: WordClock/Mqtt.cpp (hunk 5 above)
Comment (jortberends) [reply WordClock/Mqtt.cpp]: And not a member of the mqtt class

File: WordClock/Mqtt.cpp (hunk 4 above)
Comment (jortberends) [reply WordClock/Mqtt.cpp]: And not a member of the mqtt class

File: WordClock/Mqtt.cpp [hunk 9]
```diff
@@ -0,0 +1,96 @@
+#include "Mqtt.h"
+
+#include <ESP8266WiFi.h>
+#include <PubSubClient.h>
```
Comment (jortberends) [inline WordClock/Mqtt.cpp]: move these includes to the Mqtt.h file

File: WordClock/Display.cpp [hunk 10]
```diff
@@ -142,6 +142,19 @@ void Display::setBrightness(uint8_t brightness) {
   pixels.show();
 }
 
+void Display::setColor(uint8_t r, uint8_t g, uint8_t b) {
+  ClockConfig.clockColor = pixels.Color(r, g, b);
```
Comment (jortberends) [inline WordClock/Display.cpp]: Wanneer Config::save() aangeroepen wordt zal de geselecteerde kleur als 'default' opgeslagen worden in het EEPROM. Is dat wat je wilde bereiken?

Comment (harmjanr) [reply WordClock/Display.cpp]: Klopt!

**Review by jortberends — APPROVED** [review]

**Pull request merged**